field_utils/FieldUtilities.F90

```diff
@@ -303,116 +204,127 @@ subroutine FieldPow(field_out,field_in,expo,rc)
 
+      if (present(typekind)) then
+         call MAPL_InfoGetInternal(fieldBundle, key=KEY_TYPEKIND, value=typekind_str, _RC)
+         select case (typekind_str)
+         case ('R4')
+            typekind = ESMF_TYPEKIND_R4
+         case ('R8')
+            typekind = ESMF_TYPEKIND_R8
+         case ('I4')
+            typekind = ESMF_TYPEKIND_I4
+         case ('I8')
+            typekind = ESMF_TYPEKIND_I8
+         case ('LOGICAL')
+            typekind = ESMF_TYPEKIND_LOGICAL
+         case default
+            _FAIL('unsupported typekind')
+         end select
+      end if
 
-   impure elemental logical function ESMF_GeomNotEqual(geom1, geom2)
-      type(ESMF_Geom), intent(in) :: geom1, geom2
-      ESMF_GeomNotEqual = .not. (geom1 == geom2)
-   end function ESMF_GeomNotEqual
-
+       if (present(ungriddedUbound)) then
+         ungridded_info = MAPL_InfoCreateFromInternal(fieldBundle, _RC)
+         ungridded_dims =  make_ungriddedDims(ungridded_info, KEY_UNGRIDDED_DIMS, _RC)
+         bounds = ungridded_dims%get_bounds()
+
+         call MAPL_InfoGetInternal(fieldBundle, key=KEY_VLOC, value=vloc, _RC)
+         if (vloc /= 'VERTICAL_DIM_NONE') then
+            call MAPL_InfoGetInternal(fieldBundle, key=KEY_NUM_LEVELS, value=num_levels, _RC)
+            select case (vloc)
+            case ('VERTICAL_DIM_CENTER')
+               bounds = [LU_Bound(1, num_levels), bounds]
+            case ('VERTICAL_DIM_EDGE')
+               bounds = [LU_Bound(1, num_levels+1), bounds]
+            case default
+               _FAIL('unsupported vertical location')
+            end select
+         end if
 
-   subroutine FieldUpdate_from_attributes(field, unusable, geom, num_levels, typekind, units, rc)
-      type(ESMF_Field), intent(inout) :: field
-      class(KeywordEnforcer), optional, intent(in) :: unusable
-      type(ESMF_Geom), optional, intent(in) :: geom
-      integer, optional, intent(in) :: num_levels
-      type(ESMF_TypeKind_Flag), optional, intent(in) :: typekind
-      character(len=*), optional, intent(in) :: units
-      integer, optional, intent(out) :: rc
+         ungriddedUbound = bounds%upper
+      end if
 
-      integer :: status
+      _RETURN(_SUCCESS)
 
-      call FieldReallocate(field, geom=geom, typekind=typekind, num_levels=num_levels, rc=rc)
+   contains
 
-      if (present(units)) then
-         call MAPL_InfoSetInternal(field, key=KEY_UNITS, value=units, _RC)
-      end if
+      subroutine get_geom(fieldBundle, geom, rc)
+         type(ESMF_FieldBundle), intent(in) :: fieldBundle
+         type(ESMF_Geom), intent(inout) :: geom
+         integer, optional, intent(out) :: rc
 
-      _RETURN(_SUCCESS)
-         
-   end subroutine FieldUpdate_from_attributes
+         integer :: status
+         type(ESMF_GeomType_Flag) :: geomtype
+         type(ESMF_Grid) :: grid
 
+         call ESMF_FieldBundleGet(fieldBundle, geomtype=geomtype, _RC)
+         if (geomtype == ESMF_GEOMTYPE_GRID) then
```

Comment:
This looks like the geom for a FieldBundle can only be a grid. Is it a MAPL constraint that FieldBundle object can only have a Grid geometry?

Reply:
I just did not want to waste time on nonexistent use cases just yet.   I've got a PR into ESMF to generalize this  (geom optional argument) which will make the logic here much simpler.